src/domains/mapDomain.ml

```diff
@@ -163,12 +158,12 @@ struct
         dok ++ (if Range.isSimple st then dprintf "%a -> %a\n" else
                   dprintf "%a -> \n  @[%a@]\n") Domain.pretty key Range.pretty st
     in
-    let group_name a () = text (Domain.class_name a) in
+    let group_name a () = text (Domain.show_group a) in
     let pretty_group  map () = fold f map nil in
     let pretty_groups rest map =
       match (fst map) with
       | 0 ->  rest ++ pretty_group (snd map) ()
```

Comment:
If I'm not mistaken, then this `fst map` comes from `groups` as some group number. But then group number `0` is special because it's printed differently?

Previously `Basetype.Variables` didn't have group 0, but now it has and changes the output?

Reply:
Mh, yes, seems like 0 means no group. Since `group_to_enum` starts with 0, I'll change this to -1 in `Printable.Std`.

Reply:
If we want to get rid of `int`-based groups and use `deriving ord` instead, then I suppose `to_group` would have to return `group option` instead.

Reply:
Yes. However, like this it's an ok trade-off I think. We only have this `let group_to_enum () = -1` in `Printable.Std` - places that use groups can derive it. If we change `to_group` to `group option` you have the problem that users can return both `Some` and `None` which makes no sense. (And lots of `Some` in the default override case.)

Reply:
Don't need `deriving ord` to get rid of `deriving enum` since constructors are `compare`d in order of definition.
Just need a reduce that starts with the first element instead of the fold.

Reply:
Got rid of `enum` and changed it to `val to_group: (t -> group) option` which is safe and not too cumbersome to write.

Reply:
The type `(t -> group) option` seems a bit weird. Why not have it as `t -> group option`? It would both be easier to work with and would match the current possibilities of groups, i.e. you can have ungrouped keys in a map, which may also contain some grouped things. I think none of the existing cases do that and it's probably not terribly useful, but the optional function is quite unusual too.

And then in the `Std` case, we could have `type group = |` ([empty variant](https://ocaml.org/releases/4.12/htmlman/emptyvariants.html)), which has no values. That would remove the weird alternative `Std.to_group`, which would give the unit group instead of no group.

Reply:
The question is if we want to allow having ungrouped mixed with grouped.
Would look like this (`None` instead of `Some Temp`):
~~~
mapping {
  tmp -> (0)
  tmp___0 -> (0)
  tmp___1 -> (1)
  tmp___2 -> (0)
  tmp___3 -> (1)
  tmp___4 -> (Unknown int([-31,31]))
  tmp___5 -> (Unknown int([-31,31]))
  tmp___6 -> (Unknown int([-31,31]))
  Global {
    __stdinp ->
      {NULL, ?}
    __stdoutp ->
      {NULL, ?}
    __stderrp ->
      {NULL, ?}
    sys_nerr -> (Unknown int([-31,31]))
    sys_errlist ->
      (Array: {}, (Unknown int([-63,63])))
    glob1 -> (5)
    glob2 -> (9)
  }
  Local {
    i -> (1)
    j -> (0)
    k -> (1)
  }
}
~~~
No strong opinion.

> the optional function is quite unusual too

In our codebase yes. It's just the apply that's harder to understand, the definitions don't differ much.

Changed it in https://github.com/goblint/analyzer/pull/203/commits/8ad23513351c985eda799cbf2c7cc642f713cbfd

Reply:
I think it wouldn't hurt to have the possibility of mixing them, especially since it doesn't make the code any more sophisticated to have the flexibility.